tachyon/math/finite_fields/quartic_extension_field.h

```diff
@@ -0,0 +1,662 @@
+    //     (a.c0 * b.c2 + a.c1 * b.c1 + a.c2 * b.c0) * x² + a.c3 * b.c3 * q * x²
+    //     (a.c0 * b.c3 + a.c1 * b.c2 + a.c2 * b.c1 * a.c3 * b.c0) * x³
+    //   = (a.c0 * b.c0 + (a.c1 * b.c3 + a.c2 * b.c2 + a.c3 * b.c1) * q,
+    //     a.c0 * b.c1 + a.c1 * b.c0 + (a.c2 * b.c3 + a.c3 * b.c2) * q,
+    //     a.c0 * b.c2 + a.c1 * b.c1 + a.c2 * b.c0 + a.c3 * b.c3 * q,
+    //     a.c0 * b.c3 + a.c1 * b.c2 + a.c2 * b.c1 * a.c3 * b.c0)
+    // Where q is Config::kNonResidue.
+
+    // See https://eprint.iacr.org/2006/471.pdf
+    // Devegili OhEig Scott Dahab --- Multiplication and Squaring on AbstractPairing-Friendly Fields.pdf; Section 5.2
+    // clang-format on
+
+    constexpr BaseField kInv2 = *BaseField(2).Inverse();
+    constexpr BaseField kInv3 = *BaseField(3).Inverse();
+    constexpr BaseField kInv4 = *BaseField(4).Inverse();
+    constexpr BaseField kInv6 = *BaseField(6).Inverse();
+    constexpr BaseField kInv12 = *BaseField(12).Inverse();
+    constexpr BaseField kInv20 = *BaseField(20).Inverse();
+    constexpr BaseField kInv24 = *BaseField(24).Inverse();
+    constexpr BaseField kInv30 = *BaseField(30).Inverse();
+    constexpr BaseField kInv120 = *BaseField(120).Inverse();
+    constexpr BaseField kNeg5 = -BaseField(5);
+    constexpr BaseField kNegInv2 = -kInv2;
+    constexpr BaseField kNegInv3 = -kInv3;
+    constexpr BaseField kNegInv4 = -kInv4;
+    constexpr BaseField kNegInv6 = -kInv6;
+    constexpr BaseField kNegInv12 = -kInv12;
+    constexpr BaseField kNegInv24 = -kInv24;
+    constexpr BaseField kNegInv120 = -kInv120;
+
+    // h1 = 2 * a.c1
+    BaseField h1 = a.c1_.Double();
+    // h2 = 4 * a.c2
+    BaseField h2 = a.c2_.Double();
+    h2.DoubleInPlace();
+    // h3 = 8 * a.c3
+    BaseField h3 = a.c3_.Double();
+    h3.DoubleInPlace().DoubleInPlace();
+    // h4 = 2 * b.c1
+    BaseField h4 = b.c1_.Double();
+    // h5 = 4 * b.c2
+    BaseField h5 = b.c2_.Double();
+    h5.DoubleInPlace();
+    // h6 = 8 * b.c3
+    BaseField h6 = b.c3_.Double();
+    h6.DoubleInPlace().DoubleInPlace();
+
+    // v0 = a.c0 * b.c0
+    BaseField v0 = a.c0_ * b.c0_;
+    // v1 = (a.c0 + a.c1 + a.c2 + a.c3) * (b.c0 + b.c1 + b.c2 + b.c3)
+    BaseField v1 =
+        (a.c0_ + a.c1_ + a.c2_ + a.c3_) * (b.c0_ + b.c1_ + b.c2_ + b.c3_);
+    // v2 = (a.c0 - a.c1 + a.c2 - a.c3) * (b.c0 - b.c1 + b.c2 - b.c3)
+    BaseField v2 =
+        (a.c0_ - a.c1_ + a.c2_ - a.c3_) * (b.c0_ - b.c1_ + b.c2_ - b.c3_);
+    // v3 = (a.c0 + 2 * a.c1 + 4 * a.c2 + 8 * a.c3) *
+    //      (b.c0 + 2 * b.c1 + 4 * b.c2 + 8 * b.c3)
+    BaseField v3 = (a.c0_ + h1 + h2 + h3) * (b.c0_ + h4 + h5 + h6);
+    // v4 = (a.c0 - 2 * a.c1 + 4 * a.c2 - 8 * a.c3) *
+    //      (b.c0 - 2 * b.c1 + 4 * b.c2 - 8 * b.c3)
+    BaseField v4 = (a.c0_ - h1 + h2 - h3) * (b.c0_ - h4 + h5 - h6);
+    // h1 = 3 * a.c1
+    h1 += a.c1_;
+    // h2 = 9 * a.c2
+    h2.DoubleInPlace().AddInPlace(a.c2_);
+    // h3 = 27 * a.c3
```

Comment:
```suggestion
    // h3 = 18 * a.c3
```

Reply:
I think **18** is right. See also Section 5.2 in https://eprint.iacr.org/2006/471.pdf 

```c++
// h3 was 8 * a.c3
// h3 = 9 * a.c3
h3 += a.c3
// h3 = 27 * a.c3
h3 += h3.Double()
```

Reply:
Oh you mean 27? Yeah 27 seems correct. Didn't see the +=.